File: windows/tests/PhoneBridge.Pairing.Tests/ProtocolTests.cs
using System.Buffers.Binary;
using PhoneBridge.Pairing;

namespace PhoneBridge.Pairing.Tests;

[TestClass]
public sealed class ProtocolTests
{
    private static readonly byte[] Window = Enumerable.Range(0,16).Select(i=>(byte)i).ToArray();
    private static PairingSession W(string code="00123456") => PairingSession.CreateWindows(Window,code.ToCharArray());
    private static PairingSession A(string code="00123456") => PairingSession.CreateAndroid(Window,code.ToCharArray(),8273,[0x30,0]);
    private static void Exchange(PairingSession w,PairingSession a,int count=8)
    { for (int i=0;i<count;i++) { var sender=i%2==0?w:a; var receiver=i%2==0?a:w; receiver.AcceptFrame(sender.CreateNextFrame()); } }
    private static byte[] Frame(byte type,int size)
    {
        var b=new byte[9+size]; "PBP1"u8.CopyTo(b);b[4]=type; BinaryPrimitives.WriteUInt32BigEndian(b.AsSpan(5),(uint)size); return b;
    }

    [TestMethod]
    [DataRow("00123456")]
    [DataRow("00000000")]
    public void FullHandshakeBindsIdentityAndGrant(string code)
    {
        using var w=W(code);using var a=A(code);Exchange(w,a);
        using var x=w.TakeConfirmation();using var y=a.TakeConfirmation();
        CollectionAssert.AreEqual(x.CopyGrant(),y.CopyGrant());CollectionAssert.AreEqual(x.ClientId,y.ClientId);
        CollectionAssert.AreEqual(x.AttemptId,y.AttemptId);CollectionAssert.AreEqual(new byte[]{0x30,0},x.CandidateCaDer);
        Assert.AreEqual((ushort)8273,x.HttpsPort);Assert.AreEqual(PairingState.Consumed,w.State);
        var copy=x.CopyGrant();copy[0]^=255;CollectionAssert.AreNotEqual(copy,x.CopyGrant());
        x.Dispose();Assert.Throws<ObjectDisposedException>(()=>x.CopyGrant());
        Assert.Throws<PairingProtocolException>(()=>w.TakeConfirmation());
    }

    [TestMethod]
    public void WrongCodeNeverYieldsConfirmation()
    { using var w=W();using var a=A("00123457"); Assert.Throws<PairingProtocolException>(()=>Exchange(w,a)); Assert.Throws<PairingProtocolException>(()=>a.TakeConfirmation()); }

    [TestMethod]
    [DataRow(0)] [DataRow(1)] [DataRow(3)] [DataRow(6)] [DataRow(7)]
    public void EarlyResultTerminatesSession(int steps)
    { using var w=W();using var a=A();Exchange(w,a,steps);Assert.Throws<PairingProtocolException>(()=>w.TakeConfirmation());Assert.AreEqual(PairingState.Failed,w.State); }

    [TestMethod]
    public void WrongDirectionAndReuseFail()
    { using var a=A();Assert.Throws<PairingProtocolException>(()=>a.CreateNextFrame());Assert.Throws<PairingProtocolException>(()=>a.AcceptFrame(Frame(1,33))); }

    [TestMethod]
    public void CancellationTerminatesAtBoundary()
    { using var w=W();Assert.Throws<PairingProtocolException>(()=>w.CreateNextFrame(new CancellationToken(true)));Assert.AreEqual(PairingState.Failed,w.State); }

    [TestMethod]
    public void ExtraFrameAfterConfirmationInvalidatesUntakenResult()
    { using var w=W();using var a=A();Exchange(w,a);Assert.Throws<PairingProtocolException>(()=>w.AcceptFrame(Frame(0x32,32)));Assert.Throws<PairingProtocolException>(()=>w.TakeConfirmation()); }

    [TestMethod]
    public void ZeroServerPortRejected()
    { using var w=W();using var a=A();Exchange(w,a,1);var frame=a.CreateNextFrame();frame[58]=frame[59]=0;Assert.Throws<PairingProtocolException>(()=>w.AcceptFrame(frame)); }

    [TestMethod]
    [DataRow(1,33)] [DataRow(2,54)] [DataRow(2,4149)] [DataRow(17,1600)] [DataRow(18,1600)]
    [DataRow(33,800)] [DataRow(34,800)] [DataRow(49,32)] [DataRow(50,32)]
    public void EveryFrameSupportsByteFragments(int type,int size)
    {
        var frame=Frame((byte)type,size);var reader=new FrameAccumulator((byte)type);
        foreach(var b in frame) reader.Feed([b]);Assert.IsTrue(reader.IsComplete);CollectionAssert.AreEqual(frame,reader.GetFrame());
        Assert.Throws<PairingProtocolException>(()=>reader.Feed([0]));Assert.IsFalse(reader.IsComplete);
    }

    [TestMethod]
    [DataRow(0)] [DataRow(4)] [DataRow(8)] [DataRow(9)] [DataRow(20)]
    public void TruncatedFrameFailsPermanently(int length)
    { var reader=new FrameAccumulator(1);reader.Feed(Frame(1,33).AsSpan(0,length));Assert.Throws<PairingProtocolException>(()=>reader.EndOfInput());Assert.Throws<PairingProtocolException>(()=>reader.Feed([0])); }

    [TestMethod]
    [DataRow(0)] [DataRow(32)] [DataRow(34)] [DataRow(8193)] [DataRow(-1)]
    public void InvalidLengthRejectedFromHeaderBeforePayload(int size)
    { var header=Frame(1,0);BinaryPrimitives.WriteUInt32BigEndian(header.AsSpan(5),unchecked((uint)size));var reader=new FrameAccumulator(1);Assert.Throws<PairingProtocolException>(()=>reader.Feed(header)); }

    [TestMethod]
    public void MagicTypeTrailingAndDuplicateRejected()
    {
        var frame=Frame(1,33);var bad=(byte[])frame.Clone();bad[0]^=1;
        Assert.Throws<PairingProtocolException>(()=>new FrameAccumulator(1).Feed(bad));
        Assert.Throws<PairingProtocolException>(()=>new FrameAccumulator(2).Feed(frame));
        Assert.Throws<PairingProtocolException>(()=>new FrameAccumulator(1).Feed([..frame,0]));
        using var w=W();using var a=A();var hello=w.CreateNextFrame();a.AcceptFrame(hello);
        Assert.Throws<PairingProtocolException>(()=>a.AcceptFrame(hello));
    }

    [TestMethod]
    [DataRow(9)] [DataRow(10)] [DataRow(26)] [DataRow(60)] [DataRow(61)]
    public void InvalidServerHelloFails(int offset)
    { using var w=W();using var a=A();Exchange(w,a,1);var hello=a.CreateNextFrame();hello[offset]^=1;Assert.Throws<PairingProtocolException>(()=>w.AcceptFrame(hello)); }

    [TestMethod]
    [DataRow(42)] [DataRow(58)] [DataRow(62)]
    public void BoundServerHelloChangeFailsAtProof(int offset)
    { using var w=W();using var a=A();Exchange(w,a,1);var hello=a.CreateNextFrame();hello[offset]^=1;w.AcceptFrame(hello);Assert.Throws<PairingProtocolException>(()=>a.AcceptFrame(w.CreateNextFrame())); }

    [TestMethod]
    public void CrossSessionReplayAndReflectionFail()
    {
        using var w=W();using var a=A();Exchange(w,a,2);var r1=w.CreateNextFrame();a.AcceptFrame(r1);var peer=a.CreateNextFrame();
        var reflection=(byte[])r1.Clone();reflection[4]=0x12;Assert.Throws<PairingProtocolException>(()=>w.AcceptFrame(reflection));
        using var w2=W();using var a2=A();Exchange(w2,a2,2);_=w2.CreateNextFrame();Assert.Throws<PairingProtocolException>(()=>w2.AcceptFrame(peer));
    }

    [TestMethod]
    [DataRow(0)] [DataRow(1)] [DataRow(2)] [DataRow(3)]
    public void InvalidGroupOrScalarRejected(int kind)
    {
        using var w=W();using var a=A();Exchange(w,a,2);var r1=w.CreateNextFrame();
        if(kind<3) { Array.Clear(r1,393,384);if(kind==1)r1[776]=1;if(kind==2)Array.Fill(r1,(byte)255,393,384); }
        else Array.Fill(r1,(byte)255,1161,32);
        Assert.Throws<PairingProtocolException>(()=>a.AcceptFrame(r1));Assert.AreEqual(PairingState.Failed,a.State);
    }

    [TestMethod]
    [DataRow(4)] [DataRow(5)] [DataRow(6)] [DataRow(7)]
    public void ModifiedProofOrMacRejected(int step)
    {
        using var w=W();using var a=A();Exchange(w,a,step);var sender=step%2==0?w:a;var receiver=step%2==0?a:w;
        var frame=sender.CreateNextFrame();frame[^1]^=1;Assert.Throws<PairingProtocolException>(()=>receiver.AcceptFrame(frame));
    }

    [TestMethod]
    public void InvalidInputsAndClosedSessionsRejected()
    {
        Assert.Throws<PairingProtocolException>(()=>W("１２３４５６７８"));
        Assert.Throws<PairingProtocolException>(()=>PairingSession.CreateAndroid(Window,"00123456".ToCharArray(),0,[1]));
        Assert.Throws<PairingProtocolException>(()=>PairingSession.CreateAndroid(Window,"00123456".ToCharArray(),1,new byte[4097]));
        using var w=W();w.Dispose();Assert.Throws<PairingProtocolException>(()=>w.CreateNextFrame());
    }
}

File: android/pairing-core/src/test/kotlin/org/phonebridge/pairing/ProtocolTests.kt
package org.phonebridge.pairing

import org.junit.Assert.*
import org.junit.Test

class ProtocolTests {
    private val window=ByteArray(16) { it.toByte() }
    private fun w(code: String="00123456") = PairingSession.createWindows(window,code.toCharArray())
    private fun a(code: String="00123456") = PairingSession.createAndroid(window,code.toCharArray(),8273,byteArrayOf(0x30,0))
    private fun exchange(w: PairingSession,a: PairingSession,count: Int=8) {
        repeat(count) { i -> (if (i%2==0) a else w).acceptFrame((if (i%2==0) w else a).createNextFrame()) }
    }
    private fun reject(block: () -> Unit) {
        try { block(); fail("Expected protocol rejection") }
        catch (e: PairingProtocolException) { assertEquals("pairing_protocol_error",e.message); assertNull(e.cause) }
    }

    @Test fun fullHandshakeAndGrantOwnership() {
        for (code in listOf("00123456","00000000")) w(code).use { w -> a(code).use { a ->
            exchange(w,a)
            w.takeConfirmation().use { x -> a.takeConfirmation().use { y ->
                assertArrayEquals(x.copyGrant(),y.copyGrant()); assertArrayEquals(x.clientId,y.clientId)
                assertArrayEquals(x.attemptId,y.attemptId); assertArrayEquals(byteArrayOf(0x30,0),x.candidateCaDer)
                assertEquals(8273,x.httpsPort); val copy=x.copyGrant(); copy[0]=(copy[0].toInt() xor 255).toByte()
                assertFalse(copy.contentEquals(x.copyGrant())); assertEquals(PairingState.CONSUMED,w.state)
                x.close(); assertThrows(IllegalStateException::class.java) { x.copyGrant() }
                reject { w.takeConfirmation() }
            } }
        } }
    }
    @Test fun wrongCodeAndEarlyResult() {
        w().use { w -> a("00123457").use { a -> reject { exchange(w,a) }; reject { a.takeConfirmation() } } }
        for (steps in listOf(0,1,3,6,7)) w().use { w -> a().use { a ->
            exchange(w,a,steps); reject { w.takeConfirmation() }; assertEquals(PairingState.FAILED,w.state)
        } }
    }
    @Test fun directionCancellationAndReuse() {
        a().use { a -> reject { a.createNextFrame() }; reject { a.acceptFrame(PairingFrames.encode(1,ByteArray(33))) } }
        w().use { w -> w.cancel(); reject { w.createNextFrame() }; assertEquals(PairingState.FAILED,w.state) }
    }
    @Test fun allFrameTypesAcceptFragments() {
        for ((type,size) in listOf(1 to 33,2 to 54,2 to 4149,17 to 1600,18 to 1600,33 to 800,34 to 800,49 to 32,50 to 32)) {
            val frame=PairingFrames.encode(type,ByteArray(size));val reader=FrameAccumulator(type)
            frame.forEach { reader.feed(byteArrayOf(it)) }; assertTrue(reader.isComplete);assertArrayEquals(frame,reader.getFrame())
            reject { reader.feed(byteArrayOf(0)) };assertFalse(reader.isComplete)
        }
    }
    @Test fun incompleteAndOversizedFrames() {
        for (size in listOf(0,4,8,9,20)) {
            val reader=FrameAccumulator(1);reader.feed(PairingFrames.encode(1,ByteArray(33)).copyOf(size))
            reject { reader.endOfInput() };reject { reader.feed(byteArrayOf(0)) }
        }
        for (size in listOf(0,32,34,8193,-1)) {
            val header=byteArrayOf(80,66,80,49,1,0,0,0,0)
            repeat(4) { header[5+it]=(size ushr (24-8*it)).toByte() }
            reject { FrameAccumulator(1).feed(header) }
        }
    }
    @Test fun magicTypeTrailingAndDuplicate() {
        val frame=PairingFrames.encode(1,ByteArray(33));val bad=frame.copyOf();bad[0]=0
        reject { FrameAccumulator(1).feed(bad) };reject { FrameAccumulator(2).feed(frame) }
        reject { FrameAccumulator(1).feed(frame+byteArrayOf(0)) }
        w().use { w -> a().use { a -> val hello=w.createNextFrame();a.acceptFrame(hello);reject { a.acceptFrame(hello) } } }
    }
    @Test fun invalidHelloFields() {
        for (offset in listOf(9,10,26,60,61)) w().use { w -> a().use { a ->
            exchange(w,a,1); val hello=a.createNextFrame();hello[offset]=(hello[offset].toInt() xor 1).toByte()
            reject { w.acceptFrame(hello) }
        } }
    }
    @Test fun boundHelloChangesInvalidateProof() {
        for (offset in listOf(42,58,62)) w().use { w -> a().use { a ->
            exchange(w,a,1);val hello=a.createNextFrame();hello[offset]=(hello[offset].toInt() xor 1).toByte()
            w.acceptFrame(hello);reject { a.acceptFrame(w.createNextFrame()) }
        } }
    }
    @Test fun reflectionAndReplay() {
        w().use { w -> a().use { a ->
            exchange(w,a,2);val r1=w.createNextFrame();a.acceptFrame(r1);val old=a.createNextFrame()
            val reflection=r1.copyOf();reflection[4]=0x12;reject { w.acceptFrame(reflection) }
            w().use { w2 -> a().use { a2 -> exchange(w2,a2,2);w2.createNextFrame();reject { w2.acceptFrame(old) } } }
        } }
    }
    @Test fun groupAndScalarBounds() {
        for (kind in 0..3) w().use { w -> a().use { a ->
            exchange(w,a,2);val r1=w.createNextFrame()
            if (kind<3) { r1.fill(0,393,777);if(kind==1)r1[776]=1;if(kind==2)r1.fill(-1,393,777) }
            else r1.fill(-1,1161,1193)
            reject { a.acceptFrame(r1) };assertEquals(PairingState.FAILED,a.state)
        } }
    }
    @Test fun modifiedRound2AndRound3() {
        for(step in 4..7) w().use { w -> a().use { a ->
            exchange(w,a,step);val sender=if(step%2==0)w else a;val receiver=if(step%2==0)a else w
            val frame=sender.createNextFrame();frame[frame.lastIndex]=(frame.last().toInt() xor 1).toByte()
            reject { receiver.acceptFrame(frame) }
        } }
    }
    @Test fun invalidInputsAndClosedSession() {
        reject { w("１２３４５６７８") }
        reject { PairingSession.createAndroid(window,"00123456".toCharArray(),0,byteArrayOf(1)) }
        reject { PairingSession.createAndroid(window,"00123456".toCharArray(),1,ByteArray(4097)) }
        w().use { w -> w.close();reject { w.createNextFrame() } }
    }
    @Test fun hkdfRfc5869KnownAnswer() {
        val value=PairingSession.hkdf(ByteArray(22) { 0x0b },ByteArray(13) { it.toByte() },ByteArray(10) { (0xf0+it).toByte() },42)
        assertEquals("3cb25f25faacd57a90434f64d0362f2a2d2d0a90cf1a5a4c5db02d56ecc4c5bf34007208d5b887185865",value.joinToString("") { "%02x".format(it.toInt() and 255) })
    }
    @Test fun extraFrameAfterConfirmationInvalidatesUntakenResult() {
        w().use { w -> a().use { a -> exchange(w,a)
            reject { w.acceptFrame(PairingFrames.encode(0x32,ByteArray(32))) };reject { w.takeConfirmation() }
        } }
    }
    @Test fun zeroServerPortRejected() {
        w().use { w -> a().use { a -> exchange(w,a,1);val frame=a.createNextFrame();frame[58]=0;frame[59]=0
            reject { w.acceptFrame(frame) }
        } }
    }
}
